### PR TITLE
Create validated commit for MLS

### PR DESCRIPTION
**Pull request by neekolas**
## tl;dr

- Adds the commit validation logic for `xmtp_id` commits
- This is an implementation of the [MLS Commit Validation section of XIP-46](https://community.xmtp.org/t/xip-46-multi-wallet-identity/639#mls-commit-validation-29) and the code should be tested and verified against the design there

## Notable changes from the spec
- We aren't actually using a `UpdateGroupMembership` proposal and instead are just storing the `GroupMembership` as an `UnknownExtension` and the proposal that updates it is a regular `GroupContextExtensions` proposal.

## TODO
- We will need an updated set of `group_permissions` compatible with the new `ValidatedCommit`
- We are going to need some more pieces in place before we can write tests. Without the Identity struct, the group and commits would have to be painfully hand-crafted to work here.

### CODE REVIEW

**Comment by neekolas** [comment]
* **#698** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/698?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @neekolas and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: xmtp_mls/src/groups/validated_commit_v2.rs [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    pub inbox_id: String,
+    pub is_creator: bool,
+    // TODO: add is_admin support
+    // pub is_admin: bool,
+}
+
+/**
+ * A [`ValidatedCommit`] is a summary of changes coming from a MLS commit, after all of our validation rules have been applied
+ *
+ * Commit Validation Rules:
```
Comment (neekolas) [inline xmtp_mls/src/groups/validated_commit_v2.rs:100]: I tried to boil the validations in XIP-46 down to these 5 rules. They don't include the rules around permissions (who is actually allowed to add/remove members or mutate metadata), which will be handled separately.

File: xmtp_mls/src/groups/validated_commit_v2.rs [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+        client: &'client Client<ApiClient>,
+    ) -> Result<Self, CommitValidationError> {
+        // Get the actor who created the commit.
+        // Because we don't allow for multiple actors in a commit, this will error if two proposals come from different authors.
+        let actor = extract_actor(staged_commit, openmls_group)?;
+
+        // Get the expected diff of installations added and removed based on the difference between the current
+        // group membership and the new group membership.
+        // Also gets back the added and removed inbox ids from the expected diff
+        let (new_group_membership, expected_installation_diff, inboxes_added, inboxes_removed) =
```
Comment (insipx) [inline xmtp_mls/src/groups/validated_commit_v2.rs:131]: Does it make sense to create a struct for this? Could still get the individual variables with struct desugaring

File: xmtp_mls/src/groups/validated_commit_v2.rs [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+    existing_group_context: &GroupContext,
+    new_group_context: &GroupContext,
+) -> Result<(GroupMembership, InstallationDiff, Vec<Inbox>, Vec<Inbox>), CommitValidationError> {
+    let old_group_membership = extract_group_membership(existing_group_context)?;
+    let new_group_membership = extract_group_membership(new_group_context)?;
+    let membership_diff = old_group_membership.diff(&new_group_membership);
+    let added_inboxes = membership_diff
+        .added_inboxes
+        .iter()
+        .map(|inbox_id| Inbox {
```
Comment (insipx) [inline xmtp_mls/src/groups/validated_commit_v2.rs:265]: might be worth `impl From<String> for Inbox` and then this becomes `.iter().map(Into::into).collect<Vec<Inbox>>()`

**Review by insipx — APPROVED** [review]
Good job! Complicated validation checking but it's well organized and the ways it can fail are clear to me

File: xmtp_mls/src/groups/validated_commit_v2.rs [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+            extract_expected_diff(
+                conn,
+                client,
+                openmls_group.export_group_context(),
+                staged_commit.group_context(),
+            )
+            .await?;
+
+        // Get the installations actually added and removed in the commit
+        let (added_installations, removed_installations, mut credentials_to_verify) =
```
Comment (insipx) [inline xmtp_mls/src/groups/validated_commit_v2.rs:141]: same here, I think maybe adding more structs would clear up some function signatures

**Review by 37ng — APPROVED** [review]

File: xmtp_mls/src/groups/validated_commit_v2.rs (hunk 3 above)
Comment (neekolas) [reply xmtp_mls/src/groups/validated_commit_v2.rs:265]: In the next PR I add the `is_creator` flag, which requires some more data than just the Inbox ID

File: xmtp_mls/src/groups/validated_commit_v2.rs (hunk 2 above)
Comment (neekolas) [reply xmtp_mls/src/groups/validated_commit_v2.rs:131]: Ya. You're probably right. I'll change that up

**Comment by neekolas** [comment]
### Merge activity

* **May 1, 5:59 PM PDT**: @neekolas merged this pull request with [Graphite](https://app.graphite.dev/github/pr/xmtp/libxmtp/698).
